burstphoto/texture/texture.metal

```diff
@@ -15,56 +15,37 @@ kernel void add_texture(texture2d<float, access::read> in_texture [[texture(0)]]
-                                 texture2d<float, access::read_write> norm_texture [[texture(2)]],
+                                 texture2d<float, access::read> in_texture_blurred [[texture(1)]],
+                                 texture2d<float, access::read> weight_highlights_texture [[texture(2)]],
+                                 texture2d<float, access::read_write> out_texture [[texture(3)]],
+                                 texture2d<float, access::read_write> norm_texture [[texture(4)]],
                                  constant int& exposure_bias [[buffer(0)]],
                                  constant float& white_level [[buffer(1)]],
                                  constant float& black_level_mean [[buffer(2)]],
+                                 constant float& color_factor_mean [[buffer(3)]],
                                  uint2 gid [[thread_position_in_grid]]) {
        
-    // load args
-    int texture_width  = in_texture.get_width();
-    int texture_height = in_texture.get_height();
-    
     // calculate weight based on exposure bias
-    float const weight_exposure = pow(2.0f, float(exposure_bias/100.0f));
+    float weight_exposure = pow(2.0f, float(exposure_bias/100.0f));
     
     // extract pixel value
     float pixel_value = in_texture.read(gid).r;
     
-    // find the maximum intensity in a 5x5 window around the main pixel
-    float pixel_value_max = 0.0f;
-      
-    for (int dy = -2; dy <= 2; dy++) {
-        int y = gid.y + dy;
-        
-        if (0 <= y && y < texture_height) {
-            for (int dx = -2; dx <= 2; dx++) {
-                int x = gid.x + dx;
-                
-                if (0 <= x && x < texture_width) {
-                    pixel_value_max = max(pixel_value_max, in_texture.read(uint2(x, y)).r);
-                }
-            }
-        }
-    }
-    
-    pixel_value_max = (pixel_value_max-black_level_mean)*weight_exposure + black_level_mean;
-    
-    float weight_pixel_value = 1.0f;
-    
-    // this ensures that pixels of the image with lowest exposure are always added
-    if (white_level < 999999.0f) {
-        // ensure smooth blending for pixel values between 0.25 and 0.99 of the white level
-        weight_pixel_value = clamp(0.99f/0.74f-1.0f/0.74f*pixel_value_max/white_level, 0.0f, 1.0f);
-    }
+    // adapt exposure weight based on the luminosity of each pixel
+    // shadows get the exposure-dependent weight for optimal noise reduction while midtones and highlights have a reduced weight for better motion robustness
+    // between 0.25 and 1.00 of the white level (based on pixel values after exposure correction), the weight becomes 1.0
+    float const luminance = min(white_level, in_texture_blurred.read(gid).r/color_factor_mean);
+    weight_exposure = max(sqrt(weight_exposure), weight_exposure * pow(weight_exposure, -0.5f/(0.25f-black_level_mean/white_level)*(luminance-black_level_mean)/white_level));
```

Comment:
I think this line should probably be broken up a bit more, the second argument to max is quite long

Reply:
Thanks for the suggestion, will revise it.